### DISCUSSION TITLE
Is there any way to find new ActionScript 3.0 language feature, like new operator?

### DISCUSSION THREAD

**Discussion by shaucky** [airsdk/Adobe-Runtime-Support#2984] [Q&A] [answered]
I'm under the impression that some version of AIR seems to mention support `?.` operator, but i can't remember which version.
 
Is there a place where can let us find information about new ActionScript 3.0 features or infor like version support?
 
I saw Apache Royale has support of `??`、`?.` and `@""` in their Compiler 0.9.10, but i'm not sure about which AIR version support 'em or not…
 
I also tried the new search function of https://airsdk.harman.com/release_notes …

**Comment by shaucky** [comment]
Well, with more researching, i find that AIR 50 support those operator earlier… But when i test the `?.` operator in Adobe Animate 2023 with frame script, it threw me some compiler error:
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/119972730/9ba30f22-3254-4be3-a065-93d67e976413)
 
The test frame script goes here:
    var a: Number = 5;
    trace(a?.toString());
 
I guess it may because of the adjustments made during the compilation of the frame script…

**Reply by shaucky** [reply]
I still get error using this operator in the class file. Is there something wrong with my coding? i'm using the AIR 50.2.3.6…

**Comment by hydroper** [comment]
Give a look at https://helpx.adobe.com/id_id/animate/kb/get-started-with-latest-airsdk.html

I guess though it's more suitable to use Adobe Animate to produce asset SWFs and embed them with `[Embed]` in an `asconfig.json` project.

**Reply by shaucky** [reply]
Thanks.
 
I know how to use the AIR SDK from Harman, but with the latest version SDK, Animate threw compiler errors with new operator like `??`.
 
After i started the thread, i tried to use Visual Studio Code with AS3&MXML plug-in. It can be compiled smoothly with new AIR SDK and new operator. (i haven't compiled AS3 using an IDE other than Animate before…)
 
So i think that could be a problem with Animate itself, for example, the Adobe Flash team didn't consider adding operators to AS3 one day, so they built an AS3 compiler into Animate?

**Reply by ajwfrost** [reply, marked as answer]
Hi @xiaoChi888 

Yes, Animate has its own compiler for some reason.. they use the AIR SDK for the packaging up of SWF files into bundles/installers/etc, but for compiling FLA/AS3 to SWF they have their own compiler and that doesn't have the support for these new operators/features. VSCode with the AS3/MXML extension is a good alternative as long as you don't need the timeline/animation/library features of Animate!

In terms of "what's new":
https://airsdk.dev/reference/actionscript/3.0/whatsnew.html (or link at the bottom of https://airsdk.dev/reference/actionscript/3.0/)
Although I've just seen that this doesn't include the AS3 changes, the "Operators" page needs to be updated......
https://airsdk.dev/reference/actionscript/3.0/operators.html

**Reply by shaucky** [reply]
Hi Andrew,
 
I saw new classes in AIR 51 are updated in referance, but please remember to update operators in referance!

**Reply by ajwfrost** [reply]
@xiaoChi888 thanks for the reminder, I had missed those!! A couple more updates are pending from the AIR 51 release so will get those done as well as updating this "operators" page for the earlier changes!

**Comment by Jason0247** [comment]
What compiler are everyone using for development? It seems that most people are using Animate because it is indeed much more convenient to develop things with visual editing.
Are there any other better options?

**Comment by FliplineStudios** [comment]
@Jason0247 We've always used Animate to create all of our assets (vector art, MovieClips with timeline animations, etc.), but we don't use it for publishing or packaging.

We instead have Animate export SWC files of all of our visual assets (with linkage in the library), and then use those elsewhere to create the main final SWF (FlashDevelop for us).  We then do the final AIR packaging from the command line with whatever the latest AIR SDK is.  If we needed to use any of the new APIs from an AIR SDK we could could overlay that in FlashDevelop though.  FlashDevelop works fine for our workflow but it's fairly out-of-date, I think VSCode or IntelliJ IDEA are more recommended these days if you're switching to something new.

**Reply by Jason0247** [reply]
Yes, I have also been using Animate because most of my projects require animation and have tight deadlines. Using Animate makes development more convenient.
I haven't used VSCode before, like the previous flex, I feel pretty good, but I stopped updating it later and haven't used it anymore.
At present, there is indeed not a very good development environment, except for Animal

**Reply by marchbold** [reply]
Animate definitely is great for the animation editing, but you should really find a better code editing and debugging IDE as Animate is just a glorified text editor when it comes to code. 

As @FliplineStudios suggested, just create your animations in Animate, export them to a SWC and use them in your IDE. IntelliJ is our go to, however VSCode is becoming very useful as well.
